README.md
# Auto-Git

Auto-Git is an extension to push project files to git in a predefined interval without taking any action by the developer.

## Getting Started

When you install Auto-Git, it will look for a workspace/folder which has to be a git-repository. If not, the extension will be disabled automatically. So make sure you work in a git-repository. This extension will also append his own directories into the .gitignore file to prevent getting uploaded to the git. If you dont have a .gitignore file, one will be created by default.

## Prerequisites

Using Auto-Git from scratch with a fresh git repository.

Your steps are easy and also a common route to work with git: make a git repository with init, setup remote data and choose upstream branch

```
git init
git remote add origin <url>
git fetch origin
git pull origin master
git branch --set-upstream-to=origin/master
```

If you are working within another branch beside master, so do not forget to switch into it using:

```
git checkout <branch>
```

I would prefer to use the git credential.helper for remembering username and password for the remote git.

Linux:

```
git config --global credential.helper store
or
git config --global credential.helper cache
```

MacOSX:

```
git config --global credential.helper osxkeychain
```

Windows:
For Windows, there is a helper called [Git Credential Manager for Windows](https://github.com/Microsoft/Git-Credential-Manager-for-Windows) by Microsoft

```
git config --global credential.helper wincred # obsolete
```

also do not forget to setup user.name and user.email

```
git config --global user.name "Your Full Name"
git config --global user.email "your@email.com"
```

After you created a git-repository, restart Visual Studio Code or reopen the workspace/folder. The extension recognizes it as a git repository and start automatically with predefined settings.

## Settings

There will be a **.autogit** directory with logs folder and a **autogit.json** file. You can change some behavoir like interval, logging and silent-mode in the json file.
After changing any value in the json file, you should stop and start the extension to reload the configuration. This can be done by restarting VS Code or using the extension commands.

- `runOnStart`: If true, auto-git will automatically start push cycle by given interval (see below `updateInterval`) when opening workspace (which is a git-repository).
- `logging`: Creates a log file (date based) in the .autogit/logs folder. It contains all changes from last push cycle (add/remove/modified/delete). So you can always see what happened during the push cycle. Setting to false will prevent creating log files.
- `silent`: If false, every finished push cycle, a vscode notification box will show up.
- `updateInterval`: Interval is given in seconds. Defines how long a cycle takes to push added/removed/modified/deleted files to git.
- `commitMessage`: Here you can override the default auto-git commit message.
- commitMessage optional `placeholders`:

```
{ts} represents seconds (as integer) from 1970-01-01 till now (unix timestamp)

{ts.utc} represents the UTC format e.g.: Fri, 20 May 2022 20:05:12 GMT

{ts.iso} represents the ISO format e.g.: 2022-05-20T20:05:12.072Z

{ts.locale} represents the locale format with timeZone and locale e.g. for Europe/Berlin and de-DE: 20.5.2022, 22:05:12

{ts.locale.date} represents the date only e.g. for Europe/Berlin and de-DE: 20.5.2022

{ts.locale.time} represents the time only e.g. for Europe/Berlin and de-DE: 22:05:12

{ts.locale.long} represents the date and time as long format e.g. for Europe/Berlin and de-DE: Freitag, 20. Mai 2022, 22:05:12
```

## Commands

Press F1 or CTRL+Shift+P to open VS Code Command Palette.

- Auto-Git: Version // Shows current installed version in a notification box
- Auto-Git: Init // Initializes Auto-Git for your project
- Auto-Git: Start // Starts the interval and routine
- Auto-Git: Stop // Stops the interval and routine
- Auto-Git: Restart // Reloads configuration + stop + start

## Authors

- **Eray Sönmez** - _Lead Developer_ - [Zeitner GmbH](https://www.zeitnergmbh.de)

Written in TypeScript with love. This extension is a product of Zeitner GmbH.

## License

This project is licensed under the GPL v3 License - see the LICENSE.md file for details

## Thirdparty Libraries

- Simple-Git


CHANGELOG.md
# Change Log

## 1.1.4

* Preloading extension to prevent restart of vscode when working in an empty workspace without any files.
* Cleaned up unnecessary imports
* Improved performance

## 1.1.3

* Adjusted CHANGELOG.md and README.md for more readability
* Adjusted extension information for marketplace (added tags/keywords)
* Adjusted VSCode Activation Events For Extensions. Auto-Git will now only loaded from VSCode when a folder/workspace is opened.
* Standard locale is now en-US

## 1.1.2

* Added following placeholders for custom commit messages has been implemented:

```
{ts} represents seconds (as integer) from 1970-01-01 till now (unix timestamp)

{ts.utc} represents the UTC format e.g.: Fri, 20 May 2022 20:05:12 GMT

{ts.iso} represents the ISO format e.g.: 2022-05-20T20:05:12.072Z

{ts.locale} represents the locale format with timeZone and locale e.g. for Europe/Berlin and de-DE: 20.5.2022, 22:05:12

{ts.locale.date} represents the date only e.g. for Europe/Berlin and de-DE: 20.5.2022

{ts.locale.time} represents the time only e.g. for Europe/Berlin and de-DE: 22:05:12

{ts.locale.long} represents the date and time as long format e.g. for Europe/Berlin and de-DE: Freitag, 20. Mai 2022, 22:05:12
```

* Added `locale` and `timeZone` added to autogit.json configuration file and are used for the placeholders only in the custom commit message. 
  
    locale format specification: [BCP 47](https://datatracker.ietf.org/doc/html/rfc4647#section-3.3.2)

    example: de-DE (German, as used in Germany), ar-DZ (Arabic, as used in Algeria), ar-EG (Arabic, as used in Egypt), fr-VU (French, as used in Vanuatu), [Language designators with regions](https://lingohub.com/developers/supported-locales/language-designators-with-regions)

    timeZone format specification: IANA TimeZone [List_of_tz_database_time_zones](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones)

## 1.1.0

* Added Auto-Git is not loading automatically anymore when a **new** workspace is opened. You need to initialized Auto-Git first. This behavior provides more flexibility when to use Auto-Git. Initialization can only be done in a workspace, which is a git-repository. After initialization, its set up by default to run/start when this workspace is opened again. You can change this behavior by changing "runOnStart" to false in the **autogit.json** config file.

* Added runOnStart: Run Auto-Git automatically when opening a workspace (which is a git-repository) which already has been initialized for Auto-Git.

* Added commitMessage: You are now able to set your own commit message which should be used for auto-git pushes.

* Fixed Auto-Git fails to push on some machines with non-standard git configuration.

* Adjusted the README.md file and added more information, description for the configuration file. Also some formatting.


src/extension.ts
/* eslint-disable @typescript-eslint/naming-convention */
'use strict';
import * as vscode from 'vscode';
import * as path from 'path';
import * as fs from 'fs';

import simpleGit, { SimpleGit } from 'simple-git';

let myStatusBarItem: vscode.StatusBarItem;

export function activate(context: vscode.ExtensionContext) {
	let autogit = new AutoGit();
    
    let cmdversion = vscode.commands.registerCommand('autogit.version', () => {
        vscode.window.showInformationMessage('Version 1.1.4 by Eray Sönmez <dev@ray-works.de>');
    });

	let cmdinit = vscode.commands.registerCommand('autogit.init', () => {
		if(autogit.checkWorkspace() && autogit.checkGit()){
			if(!autogit.isInitialized){
				autogit.setup();
				autogit.start();
				vscode.window.showInformationMessage('Auto-Git initialized.');
			} else {
				vscode.window.showInformationMessage('Auto-Git is already initialized.');
			}
		} else {
			vscode.window.showInformationMessage('Auto-Git can only run in a workspace and git-repository.');
		}
	});
    
	let cmdstart = vscode.commands.registerCommand('autogit.start', () => {
		if(autogit.checkWorkspace() && autogit.checkGit()){
            if(autogit.isInitialized){
                if(!autogit.running){
                    autogit.start();
                    vscode.window.showInformationMessage('Auto-Git started.');
                } else {       
                    vscode.window.showInformationMessage('Auto-Git is already running.');
                }
            } else {
                vscode.window.showInformationMessage('Run `Auto-Git: Init` before `Auto-Git: Start`.');
            }
		} else {
			vscode.window.showInformationMessage('Auto-Git can only run in a workspace and git-repository.');
		}
	});

	let cmdstop = vscode.commands.registerCommand('autogit.stop', () => {
		if(autogit.checkWorkspace() && autogit.checkGit()){
			if(autogit.running){
				autogit.stop();
				vscode.window.showInformationMessage('Auto-Git stopped.');
			} else {
				vscode.window.showInformationMessage('Auto-Git is not running.');
			}
		} else {
			vscode.window.showInformationMessage('Auto-Git can only run in a workspace and git-repository.');
		}
	});

	let cmdrestart = vscode.commands.registerCommand('autogit.restart', () => {
		if(autogit.checkWorkspace() && autogit.checkGit()){
            if(autogit.running){
                autogit.stop();
                autogit.start();
                vscode.window.showInformationMessage('Auto-Git restarted.');
            } else {
                 if(autogit.isInitialized){
                    autogit.start();
                    vscode.window.showInformationMessage('Auto-Git restarted.');
                 } else {
                     vscode.window.showInformationMessage('Run `Auto-Git: Init` before `Auto-Git: Restart`.');
                 }
            }
		} else {
			vscode.window.showInformationMessage('Auto-Git can only run in a workspace and git-repository.');
		}
	});

	context.subscriptions.push(cmdinit);
	context.subscriptions.push(cmdversion);
	context.subscriptions.push(cmdstart);
	context.subscriptions.push(cmdstop);
	context.subscriptions.push(cmdrestart);

	if(autogit.isInitialized){
		var cfg = JSON.parse(fs.readFileSync(autogit.cfg, 'utf8'));
		if(cfg.runOnStart){
			autogit.start();
		}
	}

	myStatusBarItem = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left, 100);
	context.subscriptions.push(myStatusBarItem);
	myStatusBarItem.show();
}

export function deactivate() {

}

class AutoGit implements vscode.Disposable { 
    public counter: number = 0;
    private intervalId: any = null;
    public workspace!: vscode.Uri;
    public running: boolean = false;
    private homedir!: string;
    private logsdir!: string;
    public cfg!: string;
    private gitdir!: string;
    private gitcfg!: string;
	public isInitialized: boolean = false;

    constructor(){
		this.checkWorkspace();
		this.checkGit();

		try {
            fs.statSync(this.cfg);
			this.isInitialized = true;
            var userCfg: any = JSON.parse(fs.readFileSync(this.cfg, 'utf8'));
            var currentCfg: any = this.currentConfigSchema();
            if(!this.compareKeys(userCfg, currentCfg)){
                const newProperties = Object.keys(currentCfg).filter(prop => !userCfg.hasOwnProperty(prop));

                newProperties.forEach(prop => {
                    userCfg[prop] = currentCfg[prop];
                });
                
                fs.writeFileSync(this.cfg, JSON.stringify(userCfg, null, 2));
            }
        } catch (err) {
        }
	}

    dispose(): void {
        this.stop();
    }

    public currentConfigSchema(): object {
        return { 
            "runOnStart": true,
            'updateInterval': 1800,
            'logging': true, 
            'silent': true,
            "commitMessage": "--- Auto Git Commit ---",
            "locale": "en-US",
            "timeZone": "Europe/Berlin"
        };
    }

    private compareKeys(a: object, b: object): boolean {
        var aKeys = Object.keys(a).sort();
        var bKeys = Object.keys(b).sort();
        return JSON.stringify(aKeys) === JSON.stringify(bKeys);
    }

    private updateStatusBarItem(text: string): void {
        myStatusBarItem.text = text;
    }

    public start(): void {
        var cfg = JSON.parse(fs.readFileSync(this.cfg, 'utf8'));
        this.running = true;
        this.counter = cfg.updateInterval;
        this.intervalId = setInterval(() => {
            this.counter--;
            this.updateStatusBarItem("Next Auto-Git in... " + this.counter);
            if(this.counter === 0){
                const git: SimpleGit = simpleGit(this.workspace.fsPath);
                this.updateStatusBarItem("Auto-Git: Checking files...");
                git.pull();
                git.add('.'+ path.sep + '*');
                git.status().then(async (status: any) => { 
                    let changes = status.modified.length + status.created.length + status.deleted.length + status.renamed.length;
                    if(changes > 0){
                        this.updateStatusBarItem("Auto-Git: Pushing files...");

                        let options = { 
                            weekday: 'long',
                            year: 'numeric',
                            month: 'long',
                            day: 'numeric',
                            timeZone: cfg.timeZone ?? 'Europe/Berlin',
                            hour: '2-digit',
                            minute: '2-digit',
                            second: '2-digit',
                        } as const;    

                        var replacements: { [key: string]: string }  = {
                            "{ts}": (new Date().getTime() / 1000).toString(),
                            "{ts.utc}": new Date().toUTCString(),
                            "{ts.iso}": new Date().toISOString(),
                            "{ts.locale}": new Date().toLocaleString(cfg.locale ?? 'en-US', { timeZone: cfg.timeZone ?? 'Europe/Berlin' }),
                            "{ts.locale.date}": new Date().toLocaleDateString(cfg.locale ?? 'en-US', { timeZone: cfg.timeZone ?? 'Europe/Berlin' }),
                            "{ts.locale.time}": new Date().toLocaleTimeString(cfg.locale ?? 'en-US', { timeZone: cfg.timeZone ?? 'Europe/Berlin' }),
                            "{ts.locale.long}": new Date().toLocaleString(cfg.locale ?? 'en-US', options)
                        };

                        cfg.commitMessage = cfg.commitMessage.replace(/\{.+?\}/g, (key: string): string => replacements[key]);

                        await git.commit(cfg.commitMessage ?? "--- Auto-Git Commit ---");
						var remote = status.tracking.split('/')[0] ?? "origin";
						var branch = status.tracking.split('/')[1] ?? "master";
                        await git.push(remote, branch, ['-u']);

                        console.log("[Auto-Git]: Changes since last sync: modified (" + status.modified.length + ") | created (" + status.created.length + ") | deleted (" + status.deleted.length + ") | renamed: (" + status.renamed.length + ")" );
                        if(cfg.logging){
                            var date = new Date();
                            let log = "-------------------- Auto-Git Log --------------------";
                            log += "\n" + date.toString();
                            log += "\n------------------------------------------------------";

                            log += "\n";
                            log += "\nModified files:";
                            log += "\n";
                            status.modified.forEach((element: string) => {
                                log += "* " + element + "\n";
                            });

                            log += "\n";
                            log += "\nCreated files:";
                            log += "\n";
                            status.created.forEach((element: string) => {
                                log += "* " + element + "\n";
                            });

                            log += "\n";
                            log += "\nDeleted files:";
                            log += "\n";
                            status.deleted.forEach((element: string) => {
                                log += "* " + element + "\n";
                            });

                            log += "\n";
                            log += "\nRenamed files:";
                            log += "\n";
                            status.renamed.forEach((element: string) => {
                                log += "* " + element + "\n";
                            });

                            fs.writeFileSync(this.logsdir + path.sep + 'log-' + date.getDate() + '-' + date.getMonth() + '-' + date.getFullYear() + '-' + date.getHours() + '-' + date.getMinutes() + '-' + date.getSeconds() + '.txt', log);
                        }
                        if(!cfg.silent){
                            vscode.window.showInformationMessage("Auto-Git updated " + changes + " change(s)."); 
                        }

                        this.updateStatusBarItem("Auto-Git: Push done. Starting next cycle...");
                    }
                });

                clearInterval(this.intervalId);
                this.intervalId = null;
                setTimeout(() => {
                    this.start();
                }, 5000);
            }
        }, 1000);
    }

    public stop(): void {
        if(this.running){
            clearInterval(this.intervalId);
            this.intervalId = null;
            this.running = false;
            this.updateStatusBarItem("--- Auto-Git not running ---");
        }
    }

    public setup() {
        try{
            fs.statSync(this.homedir);
        } catch (err) {
            fs.mkdirSync(this.homedir);
        }

        try{
            fs.statSync(this.logsdir);
        } catch (err) {
            fs.mkdirSync(this.logsdir);
        }

        try {
            fs.statSync(this.cfg);
        } catch (err) {
            fs.writeFileSync(this.cfg, JSON.stringify(this.currentConfigSchema(), null, 2));
        }

        try{
            fs.statSync(this.workspace.fsPath.concat(path.sep + '.gitignore'));
        }catch(err){
            fs.writeFileSync(this.workspace.fsPath.concat(path.sep + '.gitignore'), '.autogit');
        }

        let gitignore = fs.readFileSync(this.workspace.fsPath.concat(path.sep + '.gitignore'));
        if(gitignore.indexOf('.autogit') === -1) {
            fs.appendFileSync(this.workspace.fsPath.concat('.gitignore'), '.autogit');
        }

		this.isInitialized = true;
    }

    public checkGit(): boolean {
        try {
            fs.statSync(this.gitdir);
            fs.statSync(this.gitcfg);
            console.log('[Auto-Git] [OK]: Workspace is a git repository.');
            return true;
        }
        catch (err) {
            console.log('[Auto-Git] [Error]: Workspace is not a git repository, disabling extension.');
            return false;
        }
    }

    public checkWorkspace(): boolean {
        try {
			if(vscode.workspace.workspaceFolders !== undefined){
				fs.statSync(vscode.workspace.workspaceFolders[0].uri.fsPath );
				console.log('[Auto-Git] [OK]: Workspace found: ' + vscode.workspace.workspaceFolders[0].uri.fsPath);
				this.workspace = vscode.workspace.workspaceFolders[0].uri;
				
				this.homedir = this.workspace.fsPath.concat(path.sep + '.autogit');
				this.logsdir = this.workspace.fsPath.concat(path.sep + '.autogit/logs');
				this.cfg = this.workspace.fsPath.concat(path.sep + '.autogit/autogit.json');
				this.gitdir = this.workspace.fsPath.concat(path.sep + '.git');
				this.gitcfg = this.workspace.fsPath.concat(path.sep + '.git/config');

				return true;
			} else {
				console.log('[Auto-Git] [Error]: No workspace found, disabling extension.');
			}
        }
        catch (err) {
            console.log('[Auto-Git] [Error]: No workspace found, disabling extension.');
		}
		return false;
    }
}

out/extension.js
/* eslint-disable @typescript-eslint/naming-convention */
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.deactivate = exports.activate = void 0;
const vscode = require("vscode");
const path = require("path");
const fs = require("fs");
const simple_git_1 = require("simple-git");
let myStatusBarItem;
function activate(context) {
    let autogit = new AutoGit();
    let cmdversion = vscode.commands.registerCommand('autogit.version', () => {
        vscode.window.showInformationMessage('Version 1.1.3 by Eray Sönmez <dev@ray-works.de>');
    });
    let cmdinit = vscode.commands.registerCommand('autogit.init', () => {
        if (autogit.checkWorkspace() && autogit.checkGit()) {
            if (!autogit.isInitialized) {
                autogit.setup();
                autogit.start();
                vscode.window.showInformationMessage('Auto-Git initialized.');
            }
            else {
                vscode.window.showInformationMessage('Auto-Git is already initialized.');
            }
        }
        else {
            vscode.window.showInformationMessage('Auto-Git can only run in a workspace and git-repository.');
        }
    });
    let cmdstart = vscode.commands.registerCommand('autogit.start', () => {
        if (autogit.checkWorkspace() && autogit.checkGit()) {
            if (autogit.isInitialized) {
                if (!autogit.running) {
                    autogit.start();
                    vscode.window.showInformationMessage('Auto-Git started.');
                }
                else {
                    vscode.window.showInformationMessage('Auto-Git is already running.');
                }
            }
            else {
                vscode.window.showInformationMessage('Run `Auto-Git: Init` before `Auto-Git: Start`.');
            }
        }
        else {
            vscode.window.showInformationMessage('Auto-Git can only run in a workspace and git-repository.');
        }
    });
    let cmdstop = vscode.commands.registerCommand('autogit.stop', () => {
        if (autogit.checkWorkspace() && autogit.checkGit()) {
            if (autogit.running) {
                autogit.stop();
                vscode.window.showInformationMessage('Auto-Git stopped.');
            }
            else {
                vscode.window.showInformationMessage('Auto-Git is not running.');
            }
        }
        else {
            vscode.window.showInformationMessage('Auto-Git can only run in a workspace and git-repository.');
        }
    });
    let cmdrestart = vscode.commands.registerCommand('autogit.restart', () => {
        if (autogit.checkWorkspace() && autogit.checkGit()) {
            if (autogit.running) {
                autogit.stop();
                autogit.start();
                vscode.window.showInformationMessage('Auto-Git restarted.');
            }
            else {
                if (autogit.isInitialized) {
                    autogit.start();
                    vscode.window.showInformationMessage('Auto-Git restarted.');
                }
                else {
                    vscode.window.showInformationMessage('Run `Auto-Git: Init` before `Auto-Git: Restart`.');
                }
            }
        }
        else {
            vscode.window.showInformationMessage('Auto-Git can only run in a workspace and git-repository.');
        }
    });
    context.subscriptions.push(cmdinit);
    context.subscriptions.push(cmdversion);
    context.subscriptions.push(cmdstart);
    context.subscriptions.push(cmdstop);
    context.subscriptions.push(cmdrestart);
    if (autogit.isInitialized) {
        var cfg = JSON.parse(fs.readFileSync(autogit.cfg, 'utf8'));
        if (cfg.runOnStart) {
            autogit.start();
        }
    }
    myStatusBarItem = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left, 100);
    context.subscriptions.push(myStatusBarItem);
    myStatusBarItem.show();
}
exports.activate = activate;
function deactivate() {
}
exports.deactivate = deactivate;
class AutoGit {
    constructor() {
        this.counter = 0;
        this.intervalId = null;
        this.running = false;
        this.isInitialized = false;
        this.checkWorkspace();
        this.checkGit();
        try {
            fs.statSync(this.cfg);
            this.isInitialized = true;
            var userCfg = JSON.parse(fs.readFileSync(this.cfg, 'utf8'));
            var currentCfg = this.currentConfigSchema();
            if (!this.compareKeys(userCfg, currentCfg)) {
                const newProperties = Object.keys(currentCfg).filter(prop => !userCfg.hasOwnProperty(prop));
                newProperties.forEach(prop => {
                    userCfg[prop] = currentCfg[prop];
                });
                fs.writeFileSync(this.cfg, JSON.stringify(userCfg, null, 2));
            }
        }
        catch (err) {
        }
    }
    dispose() {
        this.stop();
    }
    currentConfigSchema() {
        return {
            "runOnStart": true,
            'updateInterval': 1800,
            'logging': true,
            'silent': true,
            "commitMessage": "--- Auto Git Commit ---",
            "locale": "en-US",
            "timeZone": "Europe/Berlin"
        };
    }
    compareKeys(a, b) {
        var aKeys = Object.keys(a).sort();
        var bKeys = Object.keys(b).sort();
        return JSON.stringify(aKeys) === JSON.stringify(bKeys);
    }
    updateStatusBarItem(text) {
        myStatusBarItem.text = text;
    }
    start() {
        var cfg = JSON.parse(fs.readFileSync(this.cfg, 'utf8'));
        this.running = true;
        this.counter = cfg.updateInterval;
        this.intervalId = setInterval(() => {
            this.counter--;
            this.updateStatusBarItem("Next Auto-Git in... " + this.counter);
            if (this.counter === 0) {
                const git = (0, simple_git_1.default)(this.workspace.fsPath);
                this.updateStatusBarItem("Auto-Git: Checking files...");
                git.pull();
                git.add('.' + path.sep + '*');
                git.status().then(async (status) => {
                    let changes = status.modified.length + status.created.length + status.deleted.length + status.renamed.length;
                    if (changes > 0) {
                        this.updateStatusBarItem("Auto-Git: Pushing files...");
                        let options = {
                            weekday: 'long',
                            year: 'numeric',
                            month: 'long',
                            day: 'numeric',
                            timeZone: cfg.timeZone ?? 'Europe/Berlin',
                            hour: '2-digit',
                            minute: '2-digit',
                            second: '2-digit',
                        };
                        var replacements = {
                            "{ts}": (new Date().getTime() / 1000).toString(),
                            "{ts.utc}": new Date().toUTCString(),
                            "{ts.iso}": new Date().toISOString(),
                            "{ts.locale}": new Date().toLocaleString(cfg.locale ?? 'en-US', { timeZone: cfg.timeZone ?? 'Europe/Berlin' }),
                            "{ts.locale.date}": new Date().toLocaleDateString(cfg.locale ?? 'en-US', { timeZone: cfg.timeZone ?? 'Europe/Berlin' }),
                            "{ts.locale.time}": new Date().toLocaleTimeString(cfg.locale ?? 'en-US', { timeZone: cfg.timeZone ?? 'Europe/Berlin' }),
                            "{ts.locale.long}": new Date().toLocaleString(cfg.locale ?? 'en-US', options)
                        };
                        cfg.commitMessage = cfg.commitMessage.replace(/\{.+?\}/g, (key) => replacements[key]);
                        await git.commit(cfg.commitMessage ?? "--- Auto-Git Commit ---");
                        var remote = status.tracking.split('/')[0] ?? "origin";
                        var branch = status.tracking.split('/')[1] ?? "master";
                        await git.push(remote, branch, ['-u']);
                        console.log("[Auto-Git]: Changes since last sync: modified (" + status.modified.length + ") | created (" + status.created.length + ") | deleted (" + status.deleted.length + ") | renamed: (" + status.renamed.length + ")");
                        if (cfg.logging) {
                            var date = new Date();
                            let log = "-------------------- Auto-Git Log --------------------";
                            log += "\n" + date.toString();
                            log += "\n------------------------------------------------------";
                            log += "\n";
                            log += "\nModified files:";
                            log += "\n";
                            status.modified.forEach((element) => {
                                log += "* " + element + "\n";
                            });
                            log += "\n";
                            log += "\nCreated files:";
                            log += "\n";
                            status.created.forEach((element) => {
                                log += "* " + element + "\n";
                            });
                            log += "\n";
                            log += "\nDeleted files:";
                            log += "\n";
                            status.deleted.forEach((element) => {
                                log += "* " + element + "\n";
                            });
                            log += "\n";
                            log += "\nRenamed files:";
                            log += "\n";
                            status.renamed.forEach((element) => {
                                log += "* " + element + "\n";
                            });
                            fs.writeFileSync(this.logsdir + path.sep + 'log-' + date.getDate() + '-' + date.getMonth() + '-' + date.getFullYear() + '-' + date.getHours() + '-' + date.getMinutes() + '-' + date.getSeconds() + '.txt', log);
                        }
                        if (!cfg.silent) {
                            vscode.window.showInformationMessage("Auto-Git updated " + changes + " change(s).");
                        }
                        this.updateStatusBarItem("Auto-Git: Push done. Starting next cycle...");
                    }
                });
                clearInterval(this.intervalId);
                this.intervalId = null;
                setTimeout(() => {
                    this.start();
                }, 5000);
            }
        }, 1000);
    }
    stop() {
        if (this.running) {
            clearInterval(this.intervalId);
            this.intervalId = null;
            this.running = false;
            this.updateStatusBarItem("--- Auto-Git not running ---");
        }
    }
    setup() {
        try {
            fs.statSync(this.homedir);
        }
        catch (err) {
            fs.mkdirSync(this.homedir);
        }
        try {
            fs.statSync(this.logsdir);
        }
        catch (err) {
            fs.mkdirSync(this.logsdir);
        }
        try {
            fs.statSync(this.cfg);
        }
        catch (err) {
            fs.writeFileSync(this.cfg, JSON.stringify(this.currentConfigSchema(), null, 2));
        }
        try {
            fs.statSync(this.workspace.fsPath.concat(path.sep + '.gitignore'));
        }
        catch (err) {
            fs.writeFileSync(this.workspace.fsPath.concat(path.sep + '.gitignore'), '.autogit');
        }
        let gitignore = fs.readFileSync(this.workspace.fsPath.concat(path.sep + '.gitignore'));
        if (gitignore.indexOf('.autogit') === -1) {
            fs.appendFileSync(this.workspace.fsPath.concat('.gitignore'), '.autogit');
        }
        this.isInitialized = true;
    }
    checkGit() {
        try {
            fs.statSync(this.gitdir);
            fs.statSync(this.gitcfg);
            console.log('[Auto-Git] [OK]: Workspace is a git repository.');
            return true;
        }
        catch (err) {
            console.log('[Auto-Git] [Error]: Workspace is not a git repository, disabling extension.');
            return false;
        }
    }
    checkWorkspace() {
        try {
            if (vscode.workspace.workspaceFolders !== undefined) {
                fs.statSync(vscode.workspace.workspaceFolders[0].uri.fsPath);
                console.log('[Auto-Git] [OK]: Workspace found: ' + vscode.workspace.workspaceFolders[0].uri.fsPath);
                this.workspace = vscode.workspace.workspaceFolders[0].uri;
                this.homedir = this.workspace.fsPath.concat(path.sep + '.autogit');
                this.logsdir = this.workspace.fsPath.concat(path.sep + '.autogit/logs');
                this.cfg = this.workspace.fsPath.concat(path.sep + '.autogit/autogit.json');
                this.gitdir = this.workspace.fsPath.concat(path.sep + '.git');
                this.gitcfg = this.workspace.fsPath.concat(path.sep + '.git/config');
                return true;
            }
            else {
                console.log('[Auto-Git] [Error]: No workspace found, disabling extension.');
            }
        }
        catch (err) {
            console.log('[Auto-Git] [Error]: No workspace found, disabling extension.');
        }
        return false;
    }
}
//# sourceMappingURL=extension.js.map